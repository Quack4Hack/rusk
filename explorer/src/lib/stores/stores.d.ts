type AppStore = import("svelte/store").Readable<AppStoreContent> & {
  setNetwork: (value: string) => void;
};

type AppStoreContent = {
  blocksListEntries: number;
  chainInfoEntries: number;
  fetchInterval: number;
  marketDataFetchInterval: number;
  network: string;
  networks: NetworkOption[];
  transactionsListEntries: number;
};

type MarketDataStore = import("svelte/store").Readable<MarketDataStoreContent>;

type MarketDataStoreContent = {
  data: MarketData | null;
  error: Error | null;
  isLoading: boolean;
  lastUpdate: Date | null;
};

type NetworkOption = {
  label: string;
  value: string;
};
