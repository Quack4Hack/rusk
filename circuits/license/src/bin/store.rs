// This Source Code Form is subject to the terms of the Mozilla Public
// License, v. 2.0. If a copy of the MPL was not distributed with this
// file, You can obtain one at http://mozilla.org/MPL/2.0/.
//
// Copyright (c) DUSK NETWORK. All rights reserved.

use license_circuits::LicenseCircuit;
use std::io;
use storage::store_circuit;

pub fn main() -> Result<(), io::Error> {
    store_circuit::<LicenseCircuit>(Some(String::from("LicenseCircuit")))
}
