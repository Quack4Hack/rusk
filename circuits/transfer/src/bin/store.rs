// This Source Code Form is subject to the terms of the Mozilla Public
// License, v. 2.0. If a copy of the MPL was not distributed with this
// file, You can obtain one at http://mozilla.org/MPL/2.0/.
//
// Copyright (c) DUSK NETWORK. All rights reserved.

use std::io;
use storage::store_circuit;
use transfer_circuits::*;

pub fn main() -> Result<(), io::Error> {
    // store the transfer circuits
    store_circuit::<WithdrawFromTransparentCircuit>(Some(String::from(
        "WithdrawFromTransparentCircuit",
    )))?;
    store_circuit::<SendToContractTransparentCircuit>(Some(String::from(
        "SendToContractTransparentCircuit",
    )))?;
    store_circuit::<ExecuteCircuitOneTwo>(Some(String::from(
        "ExecuteCircuitOneTwo",
    )))?;
    store_circuit::<ExecuteCircuitTwoTwo>(Some(String::from(
        "ExecuteCircuitTwoTwo",
    )))?;
    store_circuit::<ExecuteCircuitThreeTwo>(Some(String::from(
        "ExecuteCircuitThreeTwo",
    )))?;
    store_circuit::<ExecuteCircuitFourTwo>(Some(String::from(
        "ExecuteCircuitFourTwo",
    )))?;
    Ok(())
}
