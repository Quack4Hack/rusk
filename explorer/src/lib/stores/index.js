export { default as appStore } from "./appStore";
export { default as marketDataStore } from "./marketDataStore";
