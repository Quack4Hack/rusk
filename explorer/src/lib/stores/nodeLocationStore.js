import { createDataStore } from "$lib/dusk/svelte-stores";
import { duskAPI } from "$lib/services";

const locationsDataStore = createDataStore(duskAPI.getNodeLocations);
locationsDataStore.getData();

/** @type {NodeLocationStore} */
export default {
  subscribe: locationsDataStore.subscribe,
};
