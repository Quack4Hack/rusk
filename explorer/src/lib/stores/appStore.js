import { get, writable } from "svelte/store";

/** @type {NetworkOption[]}*/
const networks = [
  { label: "Testnet", value: import.meta.env.VITE_DUSK_TESTNET_NODE },
  { label: "Devnet", value: import.meta.env.VITE_DUSK_DEVNET_NODE },
];

/** @type {AppStoreContent} */
const initialState = {
  fetchInterval: Number(import.meta.env.VITE_REFETCH_INTERVAL),
  network: networks[0].value,
  networks,
};

const store = writable(initialState);
const { set, subscribe } = store;

/** @param {string} network */
const setNetwork = (network) =>
  set({
    ...get(store),
    network,
  });

/** @type {AppStore} */
export default {
  setNetwork,
  subscribe,
};
