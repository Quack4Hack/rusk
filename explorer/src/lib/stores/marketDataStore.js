import { derived, get } from "svelte/store";
import { getPathIn } from "lamb";

import { createPollingDataStore } from "$lib/dusk/svelte-stores";
import { duskAPI } from "$lib/services";

import { appStore } from ".";

const fetchInterval = get(appStore).marketDataFetchInterval;
const pollingDataStore = createPollingDataStore(
  duskAPI.getMarketData,
  fetchInterval
);

/** @type {MarketDataStoreContent} */
const initialState = {
  ...get(pollingDataStore),
  lastUpdate: null,
};

const marketDataStore = derived(
  pollingDataStore,
  ($pollingDataStore, set) => {
    const current = get(marketDataStore);
    const isDataChanged = $pollingDataStore.data !== current.data;
    const isErrorChanged = $pollingDataStore.error !== current.error;
    const hasNewData = $pollingDataStore.data && isDataChanged;
    const isRecoverableError =
      $pollingDataStore.error &&
      getPathIn($pollingDataStore, "error.cause.status") !== 429;

    set({
      data: $pollingDataStore.data ?? current.data,
      error: hasNewData ? null : $pollingDataStore.error ?? current.error,
      isLoading: $pollingDataStore.isLoading,
      lastUpdate: hasNewData
        ? new Date()
        : current.data
          ? current.lastUpdate
          : null,
    });

    if (isErrorChanged && isRecoverableError) {
      pollingDataStore.start();
    }
  },
  initialState
);

function isDataStale() {
  const { error, isLoading, lastUpdate } = get(marketDataStore);

  return (
    !!lastUpdate &&
    (error !== null || (!isLoading && Date.now() > +lastUpdate + fetchInterval))
  );
}

pollingDataStore.start();

/** @type {MarketDataStore} */
export default {
  isDataStale,
  subscribe: marketDataStore.subscribe,
};
