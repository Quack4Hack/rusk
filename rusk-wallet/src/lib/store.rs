// This Source Code Form is subject to the terms of the Mozilla Public
// License, v. 2.0. If a copy of the MPL was not distributed with this
// file, You can obtain one at http://mozilla.org/MPL/2.0/.
//
// Copyright (c) DUSK NETWORK. All rights reserved.

use std::path::PathBuf;
use std::{fmt, fs};

use blake3::Hash;
use dusk_wallet_core::Store;

use crate::lib::crypto::{decrypt, encrypt};
use crate::lib::SEED_SIZE;
use crate::Error;

/// Stores all the user's settings and keystore in the file system
#[derive(Clone)]
pub struct LocalStore {
    path: PathBuf,
    seed: [u8; SEED_SIZE],
}

impl Store for LocalStore {
    type Error = Error;

    /// Retrieves the seed used to derive keys.
    fn get_seed(&self) -> Result<[u8; SEED_SIZE], Self::Error> {
        Ok(self.seed)
    }
}

impl LocalStore {
    /// Creates a new store
    pub fn new(
        path: PathBuf,
        seed: [u8; SEED_SIZE],
    ) -> Result<LocalStore, Error> {
        // create the local store
        let store = LocalStore { path, seed };

        Ok(store)
    }

    /// Loads wallet file from file
    pub fn from_file(path: PathBuf, pwd: Hash) -> Result<LocalStore, Error> {
        // basic sanity check
        if !path.is_file() {
            return Err(Error::WalletFileNotExists);
        }

        // attempt to load and decode wallet
        let bytes = fs::read(&path)?;
        let bytes = decrypt(&bytes, pwd)?;

        if bytes.len() != SEED_SIZE {
            return Err(Error::WalletFileCorrupted);
        }

        let mut seed = [0u8; SEED_SIZE];
        seed.copy_from_slice(&bytes);

        // create and return
        Ok(LocalStore { path, seed })
    }

    /// Saves wallet to a file
    pub fn save(&self, pwd: Hash) -> Result<(), Error> {
        // encrypt seed
        let enc_seed = encrypt(&self.seed, pwd)?;

        // write file
        fs::write(&self.path, enc_seed)?;
        Ok(())
    }
}

impl fmt::Debug for LocalStore {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        write!(
            f,
            "-------------------\n\
            LocalStore: {}\n\
            Seed: {:?}",
            self.path.as_os_str().to_str().unwrap(),
            self.seed
        )
    }
}
