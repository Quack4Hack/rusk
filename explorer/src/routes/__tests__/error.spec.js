import { afterEach, describe, expect, it, vi } from "vitest";
import { cleanup, render } from "@testing-library/svelte";
import { get, readable } from "svelte/store";

import * as appStores from "$app/stores";

import ErrorPage from "../+error.svelte";

describe("Error page", () => {
  const storeWithoutError = {
    ...get(appStores.page),
    error: null,
    status: 200,
  };
  const storeWithError = {
    ...storeWithoutError,
    error: new Error("some error message"),
    status: 500,
  };
  const baseOptions = { props: {}, target: document.body };

  afterEach(cleanup);

  it("should render the error page", async () => {
    const pageStoreSpy = vi
      .spyOn(appStores, "page", "get")
      .mockReturnValue(readable(storeWithError));
    const { container } = render(ErrorPage, baseOptions);

    expect(container).toMatchSnapshot();
    pageStoreSpy.mockRestore();
  });

  it("should be able to render the error page even without an error", () => {
    const pageStoreSpy = vi
      .spyOn(appStores, "page", "get")
      .mockReturnValue(readable(storeWithoutError));
    const { container } = render(ErrorPage, baseOptions);

    expect(container).toMatchSnapshot();
    pageStoreSpy.mockRestore();
  });
});
