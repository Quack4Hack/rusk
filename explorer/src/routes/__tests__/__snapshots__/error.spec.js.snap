// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Error page > should be able to render the error page even without an error 1`] = `
<body>
  <section
    class="error-message svelte-1wuvb8w"
  >
    <h2>
      Error - 
      200
    </h2>
     
    <p>
      
    </p>
  </section>
</body>
`;

exports[`Error page > should render the error page 1`] = `
<body>
  <section
    class="error-message svelte-1wuvb8w"
  >
    <h2>
      Error - 
      500
    </h2>
     
    <p>
      some error message
    </p>
  </section>
</body>
`;
